python/fusion_engine_client/analysis/data_loader.py

```diff
@@ -507,6 +510,103 @@ def get_index(self):
+               timestamps in seconds
+
+        @return A numpy array containing P1 time values (in seconds), or `nan` if the value could not be converted.
+        """
+        # Load pose messages, which contain the relationship between P1 and GPS time. Omit any NAN values from the
+        # reference timestamps.
+        #
+        # If no pose data is available, we can't convert to P1 time. We'll return nan for any values that are not
+        # already P1 time below.
+        if PoseMessage.MESSAGE_TYPE in self.data:
+            # If the caller already read in pose data, we'll use the cached messages read with whatever parameters they
+            # specified (e.g., a specific time range). That way, A) we don't read from disk multiple times, and B) for a
+            # very long log, we don't read a ton of data from disk when the caller is only interested in a short
+            # snippet. We assume the requested timestamps won't be too far out of the time range specified for the
+            # cached messages.
+            self.logger.debug('Using existing cached pose data for P1 time conversion.')
+            pose_data = self.data[PoseMessage.MESSAGE_TYPE]
+        else:
+            result = self.read(message_types=[PoseMessage], return_numpy=True)
+            pose_data = result[PoseMessage.MESSAGE_TYPE]
+
+        idx = ~np.logical_or(np.isnan(pose_data.p1_time), np.isnan(pose_data.gps_time))
+        p1_ref_sec = pose_data.p1_time[idx]
+        gps_ref_sec = pose_data.gps_time[idx]
+
+        if len(p1_ref_sec) == 0:
+            self.logger.debug('Pose data not available. Cannot convert timestamps to P1 time.')
+            p1_ref_sec = None
+            gps_ref_sec = None
+
+        # First, convert all UTC times and all timestamp objects to GPS time or P1 values in seconds.
+        def _to_gps_or_p1(value):
+            if isinstance(value, gpstime):
+                return value.gps()
+            elif isinstance(value, datetime):
+                return gpstime.fromdatetime(value).gps()
+            elif isinstance(value, Timestamp):
+                return value.seconds
+            elif assume_utc and is_gps_time(value):
+                return unix2gps(value)
+            else:
+                return value
+
+        # If the input is a numpy array and we're not assuming the timestamps are UTC, they're already floats and either
+        # GPS timestamps or already P1 timestamps. We don't need to convert anything to GPS time, so we can skip right
+        # to the P1 conversion step below.
+        if isinstance(times, np.ndarray) and not assume_utc:
+            return_scalar = False
+            time_sec = times
+        else:
+            # If the user passed in an iterable object (list, tuple numpy array, etc.), convert all entries to GPS or
+            # P1 values in seconds.
+            try:
+                iter(times)
```

Comment:
I've never put in a check like this myself to see if an object was iterable, but I guess it turns out that this may be the best way to do it - weird.

Reply:
Yeah me neither. Stack overflow said this was the way, which is admittedly surprising but seems to work as expected.